Feature stock requisition

lomamech opened the pull request:
- Forbid the deletion of stock requisition if the requisition has been
  implied in any process in Client Side
- Adding a link to depot movement receipt in case of depot movement from
  requisition
-Movement fully completed: in case all inventories of the requisition have been received by the destination
- Adding a new component to select the requisition reference
- Addition of the link from the register of requisitions to stock
  movements
- Adding a new component to select the requisition reference
- Addition of the link from the register of requisitions to stock
  movements

- Finalization of the management of the various requisition status
- Restrict the delivery of stocks by requisition to requesting departments
  and depots and to target depots

- Display requisition reference in stock exit to service
- Update and improve SQL request for getting balance for movement from
  requisition
-Improved bh-depot-select component added filter option for user-defined
repositories
- Improvement of alert messages when out of stock by requisition
- Set data for build E2E test for exit stock by requisition
- Add requisition status excessive received
- Add stock_requisition uuid for document map
- Test: set data for E2E test and integration
- Test: Reset value for many E2E requisition test

closes #4849

lomamech committed:
feature(Stock Requisition)
- Forbid the deletion of stock requisition if the requisition has been
  implied in any process in Client Side

lomamech committed:
Solve conflict with branch master

lomamech committed:
feature(Stock Requisition)
- Adding a link to depot movement receipt in case of depot movement from
  requisition

lomamech committed:
Movement fully completed: in case all inventories of the requisition have been received by the destination

lomamech committed:
feature(Stock Requisition)
- Adding a new component to select the requisition reference
- Addition of the link from the register of requisitions to stock
  movements

- Finalization of the management of the various requisition status

lomamech committed:
feature(Stock Requisition)
- Restrict the delivery of stocks by requisition to requesting departments
  and depots and to target depots

- Display requisition reference in stock exit to service
- Update and improve SQL request for getting balance for movement from
  requisition

lomamech committed:
Sanitaze code

lomamech committed:
Resolve conflit with branch Master

lomamech committed:
Solve conflict

jniles commented on test/integration/stock/requisition.js:
```diff
@@ -11,7 +11,7 @@ describe('(/stock/requisition) The Stock Assign HTTP API', () => {
   ];
 
-  // create new stock requisition on "Depot Principal" from a service
+  // create new stock requisition on "Depot Principal" from a servicedelete stock requisition
```
This looks like a typo.

lomamech committed:
Test: Update and fixe requisition integration test

lomamech committed:
Clean and sanitaze code

lomamech committed:
Test: Update data for test, define default depot

lomamech committed:
Test: Reset value

lomamech committed:
Test: Reset value stock lot

lomamech committed:
Test: stock lots in stock

lomamech committed:
Test: stock lots in stock reset data

lomamech committed:
Test: stock lots in stock reset data

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -219,6 +219,11 @@
             "RECORD_SAME":"No changes have been made to the form",
+            "REQUISITION": {
+              "CANCEL": "Cancel the inventory requisition request",
+              "DONE": "The requisition can no longer be completed, and will be considered done",
```
You don't need this comma here.  This can be written best as `The requisition can no longer be completed and will be considered done.`

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -219,6 +219,11 @@
+            "REQUISITION": {
+              "CANCEL": "Cancel the inventory requisition request",
+              "DONE": "The requisition can no longer be completed, and will be considered done",
+              "IN_PROGRESS": "Rehabilitated the requisition request"
```
What is this used for?  It doesn't make sense to me.

lomamech replied:
This message only appears when a requisition has been canceled and it needs to be rehabilitated, **rehabilitated requirements have the status In Progress**

jniles replied:
Hmm... I think we could find a better term for this, but I cannot think of one right away.

jniles commented on client/src/js/components/bhDepotSelect/bhDepotSelect.js:
```diff
@@ -5,6 +5,7 @@ angular.module('bhima.components')
     transclude  : true,
     bindings    : {
       depotUuid        : '<',
+      user             : '<',
```
I think this should be optional `<?`.

jniles commented on client/src/js/components/bhHasRequisitionVoucher/bhHasRequisitionVoucher.js:
```diff
@@ -4,8 +4,11 @@ angular.module('bhima.components')
     controller  : bhHasRequisitionVoucherController,
     bindings    : {
       message   : '<',
+      classeLabel : '<',
```
Why can't we just use the regular success/error/info class states for this component?  Why do we need to pass in the class?

lomamech replied:
I just wanted the class to be defined dynamically

jniles commented on client/src/js/components/bhRequisitionSelect.js:
```diff
@@ -0,0 +1,50 @@
+    transclude  : true,
+    bindings    : {
+      requisitionUuid : '<',
+      depotUuid : '<',
```
Are either of these optional?  Are both required?  What happens if someone passes in both a service-uuid and a depot-uuid?

lomamech replied:
In this case, I will use two variables, one to retrieve the Uuid and the other for the type of requirement depot or service

jniles commented on client/src/js/components/bhRequisitionSelect.js:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    if ($ctrl.serviceUuid) {
+      params.requestor_uuid = $ctrl.serviceUuid;
```
Since both the `depotUuid` and the `serviceUuid` map to the `requestor_uuid`, I propose you change the `bindings` to just be:
```
requestorUuid: '<?'
```

lomamech replied:
Cool

jniles commented on client/src/js/components/bhRequisitionSelect.js:
```diff
@@ -0,0 +1,50 @@
+ */
+function RequisitionSelectController(Stock, Notify) {
+  const $ctrl = this;
+  $ctrl.label = $ctrl.label || 'FORM.LABELS.REQUISITION_REFERENCE';
```
This label should be in the `$onInit` block.

jniles commented on server/controllers/stock/requisition/requisition.js:
```diff
@@ -42,6 +42,43 @@ async function getDetails(identifier) {
+    WHERE (req.quantity - IF(mouv.quantity, mouv.quantity, 0)) > 0;
+  `;
+
+  const requisition = await db.one(sqlRequisition, [uuid]);
```
You can actually run both of these at the same time:
```js
const [requisition, items] = await Promise.all([ db.one(sqlRequisition), db.exec(sql) ]);
```

jniles commented on client/src/js/components/bhServiceOrDepot/bhServiceOrDepot.html:
```diff
@@ -41,6 +41,7 @@
   <div ng-if="$ctrl.requestorType.type_key === 'depot'">
     <bh-depot-select
       depot-uuid="$ctrl.uuid"
+      user="true"
```
What does this do?  I can't see where this is used in the `bhDepotSelect` component.

lomamech replied:
````
    if ($ctrl.user) {
      options.only_user = true;
    }
````
**https://github.com/IMA-WorldHealth/bhima/blob/c6a3349a36a2009920c98f6d9320693b9e1d17d8/client/src/js/components/bhDepotSelect/bhDepotSelect.js**
it allows to filter only the repositories defined for the user

jniles commented on client/src/modules/stock/exit/exit.js:
```diff
@@ -575,10 +575,18 @@ function StockExitController(
+          document_uuid : documentUuid,
+        };
+
+        const COMPLETED_STATUS = 6;
```
Is there somewhere global that this can be defined?  Like `bhConstant` or something like that?  I'm not sure I understand why it changed from "2" to "6" since both numbers don't mean anything to me in this context.

jniles commented on client/src/modules/stock/exit/modals/findService.modal.js:
```diff
@@ -44,17 +44,32 @@ function StockFindServiceModalController(Instance, Service, Notify, Data, Stock)
+    if (Data.depot.uuid !== requisition.depot_uuid) {
+      vm.requisitionMessage = 'REQUISITION.NOT_FOR_DEPOT';
+      vm.requisitionLabel = 'label label-warning';
+      throw new Error('The requisition is not for depots');
```
I think this might be better written "This requisition is not for this depot".

jniles reviewed (commented):
I haven't given this a proper test, but I've added some comments to try and tighten up the client-side code.

lomamech committed:
Make property user in component DepotSelect Optional

mbayopanda reviewed (commented):
Great job @lomamech I have just some comment about the use of.

- When out of stock via a requisition there is the status "Partially" which is assigned to the requisition; so that this status cannot lead to confusion can we have:
   - Partially sent
   - Partially received
   - Or just partially processed
  
- When leaving stock to a depot, the use of the requisition depends on the depot selected in the modal, is it intentional or not? otherwise, the deposit selection zone must be reset when choosing the requisition option

- The list of items not available for output via a requisition does not show any item just empty bullets.

lomamech committed:
Refactor: remove property depotUuid and serviceUuid from component bhRequisitionSelect by the using of requestorUuid

lomamech committed:
improvement(Stock Requisition)
- Define label for requisition select in $onInit
- Improve request for requisition riun both
- Use bhConstant for value of requisition complete status

lomamech committed:
Display list of items not available for output via a requisition

mbayopanda commented on client/src/i18n/en/form.json:
```diff
@@ -823,12 +829,13 @@
+              "DRAFT":"Draft",
+              "EXCESSIVE_RECEIVED_QUANTITY": "Received but with excessive quantity",
+              "IN_PROGRESS":"In Progress",
+              "PARTIALLY":"Partially"
```
Is it possible to a bit explicit, for example: Partially executed or partially processed because when we read the status column and we see just "Partially" we would like to know partially what?

jniles replied:
I agree with this.  It isn't blocking as we have had a few discussion about this so far and haven't settled on a good idea yet.

mbayopanda commented on client/src/i18n/fr/form.json:
```diff
@@ -219,6 +219,11 @@
+            "REQUISITION": {
+              "CANCEL": "Annuler la demande de réquisition des stocks",
+              "DONE": "La requisition ne pourra plus être completée et sera consideré comme fait",
+              "IN_PROGRESS": "Rehabilité la demande de requisition"
```
"IN_PROGRESS" is not "La requisition n'a pas encore été traitée" ?

mbayopanda replied:
👍 Gotcha... It would be intuitive if knew that it is for the description of requisition status.
```
"STATUS_IN_PROGRESS_DESCRIPTION":"Rehabilité la demande de requisition"
```

jniles replied:
Looks like this was addressed. 👍

mbayopanda commented on client/src/modules/stock/exit/modals/findDepot.modal.js:
```diff
@@ -17,7 +17,7 @@ function StockFindDepotModalController(Instance, Depot, Notify, Data, Stock) {
   vm.cancel = cancel;
 
-  Depot.read()
+  Depot.read(null, { only_user : true })
```
Having this restriction here will oblige to distribute stocks only to depots that the user manages. The user must have the possibility to distribute stock also to depot that he doesn't manage.

lomamech replied:
My fear is that in the context where the health zones will use BHIMA to manage stocks, the BCZs would risk having the possibility of having the possibility of distributing products (By mistake) to the depots that do not concern them

jniles replied:
This doesn't really make sense though.  Because then each CDR will need to have access to every BCZ depot.  Are you sure that makes sense?

Additionally, to send a requisition to the BCZ, the CS will need to have access to change the stock in the BCZ...

mbayopanda commented on server/controllers/stock/requisition/requisition.js:
```diff
@@ -155,7 +196,85 @@ exports.update = async (req, res, next) => {
+
+      if (movementStatus.numberInventoryPartial > 0) {
+        // Partially
+        requisition.status_id = 3;
```
Is it possible to use named constant, such as: 
```js
const PARTIALLY_PROCESSED = 3
...
requisition.status_id = PARTIALLY_PROCESSED;
```

lomamech replied:
:white_check_mark:

mbayopanda commented on server/controllers/stock/requisition/requisition.js:
```diff
@@ -155,7 +196,85 @@ exports.update = async (req, res, next) => {
+
+      if (movementStatus.numberInventoryPartial === 0 && excessiveStatus.numberExcessive > 0) {
+        // Excessive
+        requisition.status_id = 7;
```
Use constant also

lomamech replied:
:white_check_mark:

mbayopanda reviewed (commented):
@lomamech once again this is great work 👍 

I have noticed some stuff to enhance the work so I commented them

lomamech committed:
Removal of the restriction on the exit of stocks at the depots for the deposit defined to the user, for a possible update of the rights of the users at the depots

lomamech committed:
Use constant for define requisition status partial and excessive

lomamech committed:
Fix and Resolve conflict

lomamech committed:
Fix and resolve conflict

jniles commented on client/src/js/components/bhDepotSelect/bhDepotSelect.js:
```diff
@@ -5,6 +5,7 @@ angular.module('bhima.components')
     transclude  : true,
     bindings    : {
       depotUuid        : '<',
+      user             : '<?',
```
We could probably just make this `@?` instead of `<?` to save a listener.

I would also rename it to describe what it does.  It does not take a user object.   It filters the depots on the user's permissions.  So something like `filterByUserPermissions` or `filterOnUser` or something would make more sense to me.

jniles commented on client/src/js/components/bhHasRequisitionVoucher/bhHasRequisitionVoucher.html:
```diff
@@ -2,22 +2,24 @@
 
     <div class="help-block" ng-show="$ctrl.message">
-      <p translate>{{ $ctrl.message }}</p>
+      <p class="{{ $ctrl.classeLabel }}" translate>{{ $ctrl.message }}</p>
```
This should probably use [`ng-class`](https://docs.angularjs.org/api/ng/directive/ngClass) here.

jniles commented on client/src/modules/stock/exit/modals/findService.modal.js:
```diff
@@ -44,17 +44,32 @@ function StockFindServiceModalController(Instance, Service, Notify, Data, Stock)
 
-    if (requisition.status_key === 'done') {
+    if (requisition.status_key === 'done' || requisition.status_key === 'completed'
+      || requisition.status_key === 'excessive') {
```
It looks like you have this logic in two places - `findService.modal.js` and `findDepot.modal.js`.  They look really similar - in fact, you throw the error "not for this depot" here, even though this is the find service modal

I think a good next step would be to refactor this into a `RequisitionService.validateRequsitionResult()` and use that in both modals.

lomamech replied:
:+1:

jniles commented on client/src/modules/stock/movements/templates/action.cell.html:
```diff
@@ -18,6 +18,12 @@
     </li>
 
+    <li ng-if="row.entity.stock_requisition_uuid">
+      <a data-method="view-requisition" href ui-sref="stockRequisition({ filters : [{ key : 'reference', value : row.entity.document_requisition, displayValue: displayValue}]})">
```
We should probably put "cacheable: false" here.

jniles replied:
Also, we might want to add the `period` filter with a value 'allTime' to ensure it shows all time.

jniles replied:
Also, we should be filtering on the UUID of the requisition, not on the text, I think.

jniles commented on client/src/modules/stock/movements/templates/action.cell.html:
```diff
@@ -18,6 +18,12 @@
 
+    <li ng-if="row.entity.stock_requisition_uuid">
+      <a data-method="view-requisition" href ui-sref="stockRequisition({ filters : [{ key : 'reference', value : row.entity.document_requisition, displayValue: displayValue}]})">
+        <i class="fa fa-file-pdf-o"></i> <span translate>REPORT.VIEW_REQUISITION</span>
```
Does this open a PDF?  If not, we probably should not show the PDF icon.

jniles commented on client/src/modules/stock/requisition/modals/status.tmpl.html:
```diff
@@ -0,0 +1,54 @@
+</div>
+
+<form 
+  name="ModalForm"
```
Why does this modal exist?  Shouldn't these be modified automatically?  Except for canceled which is an action, not a status.

lomamech replied:
The modals can also display a small message, to remind the user of the scope of each action
![image](https://user-images.githubusercontent.com/8035588/94549443-249c7180-024a-11eb-9308-f2ecedf6d126.png)

jniles replied:
Personally, I find having one radio button confusing. 

I think a better interface would use ModalAction for this - where the message is just displayed and the user chooses ok/cancel.  This would involve the user clicking "cancel", "confirm", etc on a dropdown menu in the 'actions' column first though.

jniles commented on client/src/modules/stock/requisition/modals/action.modal.html:
```diff
@@ -50,6 +56,7 @@
     <div id="depot-supplier" class="form-group">
       <bh-depot-select
         depot-uuid="$ctrl.model.depot_uuid"
+        user="true"
```
This appears buggy:

![image](https://user-images.githubusercontent.com/896472/94270885-b77e9880-ff38-11ea-8974-392ecdc64c67.png)

jniles commented on server/controllers/inventory/depots/index.js:
```diff
@@ -208,12 +213,20 @@ function searchByName(req, res, next) {
       LEFT JOIN sector s ON s.uuid = v.sector_uuid
       LEFT JOIN province p ON p.uuid = s.province_uuid
       LEFT JOIN country c ON c.uuid = p.country_uuid
+      LEFT JOIN depot_permission dp  ON dp.depot_uuid = d.uuid
```
I think you need to remove these.  There isn't a need do a left join here and only causes repeats in this list.

![image](https://user-images.githubusercontent.com/896472/94271058-fca2ca80-ff38-11ea-99e1-8dc02a41d4c3.png)

jniles commented on server/controllers/stock/core.js:
```diff
@@ -108,6 +109,7 @@ function getLotFilters(parameters) {
   filters.equals('invoice_uuid', 'invoice_uuid', 'm');
   filters.equals('purchase_uuid', 'origin_uuid', 'l');
   filters.equals('tag_uuid', 'tags', 't');
+  filters.equals('stock_requisition_uuid', 'uuid', 'sr');
```
Let's make this a custom filter that looks up the value in requisition movements similar to how the `only_user` works:

```js

  filters.custom(
    'user_id',
    'd.uuid IN (SELECT depot_permission.depot_uuid FROM depot_permission WHERE depot_permission.user_id = ?)',
  );

```

lomamech replied:
here I did not understand very well

jniles commented on server/controllers/stock/core.js:
```diff
@@ -345,14 +347,18 @@ async function getMovements(depotUuid, params) {
+  LEFT JOIN service AS serv ON serv.uuid = m.entity_uuid
+  LEFT JOIN stock_requisition_movement smr ON smr.document_uuid = m.document_uuid
+  LEFT JOIN stock_requisition sr ON sr.uuid = smr.stock_requisition_uuid
+  LEFT JOIN document_map sr_m ON sr_m.uuid = sr.uuid
```
You are adding a huge amount of JOINs to a query that is executed frequently!  Do we really need the requisition uuid with every movement, even if the depot rarely uses requisitions...?

I'm not really comfortable with this change.  Where is it that you need the stock requisition text?  Can you just look that up in a separate query?

lomamech replied:
It was just because of this constraint where there was almost a link between the stock movements and the requirements.

![image](https://user-images.githubusercontent.com/8035588/94550735-25360780-024c-11eb-862a-960d8f321046.png)

jniles replied:
Yes, but this doesn't mean there needs to be a JOIN.  This just means there needs to be a link.

lomamech replied:
I propose to add the requisition_uuid column on the stock_movement table in the same way as there is the invoice_uuid column
![image](https://user-images.githubusercontent.com/8035588/94658779-2a05c480-02fb-11eb-9220-32c15379e4a6.png)

jniles commented on server/controllers/stock/requisition/requisition.js:
```diff
@@ -155,7 +199,85 @@ exports.update = async (req, res, next) => {
+              SELECT BUID(i.uuid) inventory_uuid, i.code, it.text as inventoryType, sri.quantity
+              FROM stock_requisition_item sri
+              JOIN inventory i ON i.uuid = sri.inventory_uuid
+              JOIN inventory_type it ON i.type_id = it.id
```
You don't need the JOINs on the inventory_type.  In fact, if all you need is the movement quantity, you probably don't need a lot of these JOINs.  Can you try to remove them?

jniles commented on server/controllers/stock/requisition/requisition.js:
```diff
@@ -155,7 +199,85 @@ exports.update = async (req, res, next) => {
+              SELECT BUID(i.uuid) inventory_uuid, i.code, it.text as inventoryType, sri.quantity
+              FROM stock_requisition_item sri
+              JOIN inventory i ON i.uuid = sri.inventory_uuid
+              JOIN inventory_type it ON i.type_id = it.id
```
Same comment about the inventory_type here.  It doesn't look like you need it.

lomamech replied:
:+1:

jniles commented on server/models/admin.sql:
```diff
@@ -63,6 +63,11 @@ BEGIN
 
+  -- stock_requisition
+  INSERT INTO document_map
+    SELECT stock_requisition.uuid, CONCAT_WS('.', 'SREQ', stock_requisition.reference)
```
Normally, our document maps are composed of three different elements, but here you only use two.  Can you add in the project to make it into three?

Ex) Invoices -> `IV.HBB.123`
Ex) Stock Movements -> `SM.12.1234`

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -777,13 +777,27 @@ INSERT INTO `actions`(`id`, `description`) VALUES
+DROP TABLE IF EXISTS `stock_requisition_movement`;
+CREATE TABLE `stock_requisition_movement` (
+  `stock_requisition_uuid` BINARY(16) NOT NULL,
+  `document_uuid`BINARY(16) NOT NULL,
```
I think you need a space between the document_uuid and the BINARY here.

jniles reviewed (commented):
Most of this is looks good.

I'm actually quite concerned about all the `JOIN`s added here.   We are adding a huge amount of overhead to frequently called routes.  Our stock routes are already slow (Kintambo was upset with us when we used a raspberry pi there).  I think we need to find a better way than introducing more JOINs into the stock routes.

I've made some proposals, but you know where you need them more than I do.

mbayopanda reviewed (changes requested):
In the requisition form, the supplier depot must not be restricted to the user permission on the depot

lomamech committed:
feature(Stock Requisition)
- Update label be explicit Partially delivered for requisition

lomamech committed:
Improve component bhDepotSelect rename params for select depot by User

lomamech committed:
improvement(Stock Requisition)
- Use ng-class instead bh-requistion-select
- Add property cacheable false in link from stock movement to stock
  requisition, and add filter allTime, and filter requisition by UUID

lomamech committed:
improvement(Stock Requisition)
- remplace icon for PDF in link from stock movement to requision by icon
  for file-text
- Remove not necessary link from inventory_type from checkExcessive balance

lomamech committed:
Improvement: Remove not necessary link from inventory_type from check
requition balance

lomamech committed:
Resolve conflict

lomamech committed:
Add property project_id in stock requisition

lomamech committed:
refactor(Stock Requisition)
- Add Projet abbr in document maps for requisition
- sanitaze code

lomamech committed:
improve(Triggers)
- Define reference for requisition by project

lomamech committed:
Finalize and Update Migrate file: featureStockRequisition

lomamech committed:
improvement(Stock Requisition)
- Refactor Stock exit from requisition add column stock_requisition_uuid
  in table stock_mouvement instead to use a join table

- Delete table stock_requisition_movement
- Refactor all link to filter requisition from stock_movement instead
  stock movement for requisition

lomamech committed:
Test: remove value

jniles commented on client/src/modules/stock/requisition/modals/action.modal.html:
```diff
@@ -50,6 +56,7 @@
     <div id="depot-supplier" class="form-group">
       <bh-depot-select
         depot-uuid="$ctrl.model.depot_uuid"
+        filter-by-user-permission="true"
```
Nice 👍

jniles commented on client/src/modules/stock/requisition/templates/action.cell.html:
```diff
@@ -1,20 +1,42 @@
-<div class="ui-grid-cell-contents text-right" uib-dropdown dropdown-append-to-body uib-dropdown-toggle ng-if="!row.groupHeader">
-  <a href>
+<div class="ui-grid-cell-contents text-right" uib-dropdown dropdown-append-to-body>
+  <a href uib-dropdown-toggle>
```
Nice 👍

jniles commented on client/src/modules/stock/requisition/templates/action.cell.html:
```diff
@@ -1,20 +1,42 @@
   </a>
 
-  <ul data-action="{{ rowRenderIndex}}"  class="dropdown-menu" bh-dropdown-menu-auto-dropup uib-dropdown-menu>
+  <ul class="dropdown-menu-right" bh-dropdown-menu-auto-dropup uib-dropdown-menu>
```
Nice 👍

jniles commented on server/controllers/stock/requisition/requisition.js:
```diff
@@ -78,11 +119,13 @@ function getStockRequisition(params) {
   filters.equals('depot_uuid', 'depot_uuid', 'sr');
   filters.equals('requestor_uuid', 'requestor_uuid', 'sr');
   filters.equals('user_id', 'user_id', 'sr');
+  filters.equals('project_id', 'project_id', 'sr');
```
Nice 👍

jniles commented on server/models/schema.sql:
```diff
@@ -1936,15 +1936,19 @@ CREATE TABLE `stock_requisition` (
   `updated_at`          TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
   `created_at`          TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
-  PRIMARY KEY (`reference`),
+  PRIMARY KEY (`uuid`),
```
👍

jniles commented on server/models/schema.sql:
```diff
@@ -1958,22 +1962,23 @@ CREATE TABLE `stock_requisition_item` (
+  `user_id`                 SMALLINT(5) UNSIGNED NOT NULL,
+  `reference`               INT(11) UNSIGNED NOT NULL,
+  `invoice_uuid`            BINARY(16) NULL,
+  `stock_requisition_uuid`  BINARY(16) NULL,
```
In the long run, we could probably merge the columns for `invoice_uuid` and _`stock_requisition_uuid`.

jniles reviewed (approved):
This looks good to me.  All my previous comments have been addressed, and then some.  

Great work @lomamech!

bors r+

lomamech committed:
Fix and resolve conclict

lomamech committed:
Test: Update stock inventories
